### ISSUE TITLE
LessThanLength for SameLen

### ISSUE THREAD

**Issue by as56** [kelloggm/checker-framework#88]
consider:
```
import org.checkerframework.checker.index.qual.*;

public class SameLenBug {

	public static void bug3() {
		int[] b = new int[5];
		int[] a = new int[b.length*2];
		for( int i = 0; i < b.length; i++ )
            		a[ i ] = b[ i ]; //checker warning: @LTLOf("a") required
			
	}
}
```
Note that array b is twice the size of array a, therefore an index for b should also be an index for a

warning produced by the checker:
SameLenBug.java:21: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
            		a[ i ] = b[ i ]; //checker warning: @LTLOf("a") required
            		   ^
  found   : @LTLengthOf("b") int
  required: an integer less than a's length (@LTLengthOf("a"))

**Comment by mernst** [comment]
I cannot reproduce this error.  I checked in the following test case, as file `checker/tests/samelen/LessThanLen.java`, and the tests passed.

Please reopen this if I have made a mistake.

```java
import org.checkerframework.checker.index.qual.*;

public class LessThanLen {

    public static void m1() {
        int[] shorter = new int[5];
        int[] longer = new int[shorter.length * 2];
        for (int i = 0; i < shorter.length; i++) {
            longer[i] = shorter[i];
        }
    }

    public static void m2(int[] shorter) {
        int[] longer = new int[shorter.length * 2];
        for (int i = 0; i < shorter.length; i++) {
            longer[i] = shorter[i];
        }
    }

    public static void m3(int[] shorter) {
        int[] longer = new int[shorter.length + 1];
        for (int i = 0; i < shorter.length; i++) {
            longer[i] = shorter[i];
        }
    }

}
```

**Comment by kelloggm** [comment]
The reported issue is an incorrect warning issued by the upperbound checker. While the SameLen checker could (should) prevent such a warning from being issued, it cannot issue the warning itself, so the test will "pass" (i.e. issue no warnings) if run as a SameLen test. The test case should be moved to upperbound (where it will fail; I checked). I will resolve this issue.

**Comment by mernst** [comment]
Why are there separate directories for all the different subparts of the
Index Checker, which are implementation details? It could be simpler, and
catch more errors, to put them all in a directory for the whole Index
Checker.

**Comment by kelloggm** [comment]
My intention is to keep the pieces separate, for two reasons:

1. to enable re-use. The lowerbound checker, for instance, might be
valuable in other contexts than just array indices.

2. To prevent the high degree of coupling seen in the original index
checker. The pieces are separate systems, and the intent is to preserve the
boundary between them.

In summary, I think maintaining the boundary is worth it, because it
prevents the pieces from becoming too interdependent. I think this is true
even if sometimes we have problems like this. I do think we should have an
index test directory, though, which runs the full index checker on its
tests.

**Comment by mernst** [comment]
I agree that keeping the parts of the type-checker implementation separate is a good thing.  I was asking why there are separate *testing* directories for all the different subparts of the Index Checker; sorry if that was unclear. 

The case for separating the tests seems less compelling to me.  It reduces the amount of testing done and requires people to remember different commands to run for different test cases.  What's an example of a test that *shouldn't* be run for certain parts of the Index Checker, and cannot be addressed with a small change to the code or with `@SuppressWarnings`?

**Comment by kelloggm** [comment]
I don't know of such a test. I think a combined testing directory, while
keeping the implementations separate, is a good idea.

**Comment by mernst** [comment]
OK, sounds great.  Thanks!

**Comment by mernst** [comment]
Suzanne's (not yet checked in) code handles the third example (`m3`) in my test case (turning it into an offset), and a small change can handle the first two examples, refining `shorter.length` to `@LTEqLengthOf("longer")`.

**Comment by smillst** [comment]
Fixed via typetools/checker-framework#1070.
